Ingest startd ads into ES

jvansanten opened the pull request.

jvansanten committed:
Ingest startd ads into ES

dsschult commented on condor_utils.py:
```diff
@@ -721,3 +721,74 @@ def read_from_collector(address, history=False):
+            htcondor.AdTypes.Startd,
+            (
+                "EnteredCurrentState>={}"
+                .format(start_stamp)
```
If a slot has been running a job for 9 hours, but our look back time is 8 hours, do we really not want to see it?  Maybe `LastHeardFrom` is better for weeding out stale startds.

jvansanten replied:
The idea here was not to weed out stale startds, but to select only those whose state has changed since the last run. It's true that this would miss startds running long jobs that are killed by the parent batch system between runs. I'll change this.

jvansanten committed:
Use daemon start time in index

Glidein slot names like slot1@26284@kepler12.zeuthen.desy.de use PIDs to
distinguish between startds running on the same node. These may roll over after
a time, causing collisions with future glideins.

jvansanten committed:
Normalize GPUs in machine ads

dsschult commented on condor_utils.py:
```diff
@@ -762,3 +767,82 @@ def read_from_collector(address, history=False):
+            (
+                "EnteredCurrentState>={}"
+                .format(start_stamp)
+            ),
```
One additional thing to consider: do you want to see all slots, only the main partitionable slot, or just the dynamic slots that have jobs in them?

jvansanten replied:
I originally wanted to track partitioned slots explicitly, but that's probably a bad idea. I'll just reject `Dynamic` slots.

dsschult reviewed (commented):
A general comment: this won't produce pretty graphs in grafana, because it overwrites previous entries for the same startd while alive.  But single entries for each startd is probably easier for scripted analytics.

jvansanten committed:
Add Mainz_MogonI to site list

jvansanten committed:
Extract site names from ips or domain names

jvansanten committed:
Only consume partitionable slots

jvansanten commented:
I would prefer to have single entries for each startd for later matching to jobs. It may be possible to make acceptable graphs for glideins by charging their contribution to the time bin where they were last updated before exiting. This will not handle persistent glideins like the ones Claudio runs, but maybe that's okay.

Alternatively, one could obtain the number of glideins active in each time bin by subtracting a cumulative histogram of LastHeardFrom from a cumulative histogram of DaemonStartTime. I've gotten as far as making those two histograms in a freestanding ES query, but subtracting them is currently beyond my skill.

dsschult commented:
I'm probably going to go the approach of just making two scripts.  One here for your queries, and another for pretty graphs.

dsschult commented:
I'm ok to merge it now, if you're done making changes.

jvansanten commented:
I'm done for now. Let's collect metrics for a few days and see if it's useful.

jvansanten commented:
> I'm probably going to go the approach of just making two scripts. One here for your queries, and another for pretty graphs.

More things learned: [This note](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-range-field-note.html) seems to indicate that Elastic added support for `date_histogram`s on `date_range` fields at some point. This is probably the actual solution to the problem of representing how many of a thing were in some state at a given time, rather than just the start or end times.

Given that I can't get the example to work with 6.8, I suppose it was after that. The underlying index support [has been in Lucene for years now](https://www.elastic.co/blog/numeric-and-date-ranges-in-elasticsearch-just-another-brick-in-the-wall), so I guess it's just that aggregations were not supported until remarkably recently.
